quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,885 @@
+    ///
+    /// Takes a previously locked guard for easier mutext management.
+    async fn save_state<'a>(
+        &self,
+        wrapped_cache: &MutexGuard<'_, WrappedCache>,
+    ) -> StorageResult<()> {
+        let mut cache_items = vec![];
+        for (path, size) in wrapped_cache.lru.iter() {
+            cache_items.push((path.clone(), *size));
+        }
+
+        let cache_state = CacheState {
+            local_storage_uri: self.local_storage.uri(),
+            disk_capacity: self.disk_capacity,
+            items: cache_items,
+        };
+
+        let file_path = self.local_storage_root.join(CACHE_STATE_FILE_NAME);
+        let content: Vec<u8> = serde_json::to_vec(&cache_state)
+            .map_err(|err| StorageErrorKind::InternalError.with_error(err))?;
+        atomic_write(&file_path, &content)?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Storage for StorageWithUploadCache {
+    async fn put(&self, path: &Path, payload: PutPayload) -> StorageResult<()> {
+        self.remote_storage.put(path, payload.clone()).await?;
+
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        let payload_length = payload.len().await? as usize;
+        if self.exceeds_capacity(payload_length, locked_disk_cache.num_files + 1) {
+            return Ok(());
+        }
+
+        if let Some(item_num_bytes) = locked_disk_cache.lru.pop(&path.to_path_buf()) {
+            if self.local_storage.delete(path).await.is_err() {
+                return Ok(());
+            }
+            locked_disk_cache.num_bytes -= item_num_bytes;
+            locked_disk_cache.num_files -= 1;
+        }
+
+        while self.exceeds_capacity(
+            locked_disk_cache.num_bytes + payload_length,
+            locked_disk_cache.num_files + 1,
+        ) {
+            if let Some((item_path, item_num_bytes)) = locked_disk_cache.lru.pop_lru() {
+                if self
+                    .local_storage
+                    .delete(item_path.as_path())
+                    .await
+                    .is_err()
+                {
+                    return Ok(());
+                }
+                locked_disk_cache.num_bytes -= item_num_bytes;
+                locked_disk_cache.num_files -= 1;
+            }
+        }
+
+        self.local_storage.put(path, payload).await?;
+        locked_disk_cache.num_bytes += payload_length;
+        locked_disk_cache.num_files += 1;
+        locked_disk_cache
+            .lru
+            .put(path.to_path_buf(), payload_length);
+        let _ = self.save_state(&locked_disk_cache).await;
+        Ok(())
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<()> {
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        if locked_disk_cache.lru.get(&path.to_path_buf()).is_none() {
+            return self.remote_storage.copy_to_file(path, output_path).await;
+        }
+
+        let _ = self.save_state(&locked_disk_cache).await;
```

Comment:
why do we need to save the state here?

Reply:
same remark for `get_slice`, `get_all`

Reply:
The lru internal structure has moved so is the list of items ordering.
But we can also leave out this and only save when something is added or evicted.

Reply:
We need to clarify the usage of this cache layer. What is it for? It would be cool to explain that in the docs of `StorageWithUploadCache`
From my understanding, this is essentially to avoid a costly `copy_to_file`.  We don't want to hit the cache for other operations except for put. Right?

Reply:
If you agree, we should never touch the cache except for `put` and `copy_to_file`.